### DISCUSSION TITLE
Vensim MCMC

### DISCUSSION THREAD

**Discussion by hyunjimoon** [Data4DM/BayesSD#96] [people; relating]
Time delay exists till HMC, so this summarize all comments for current Vensim MCMC
- psrf jumps right after burnin (because it resets)?

**Comment by hyunjimoon** [comment]
For hierarchical model, none of the team (city, world, service) perfectly recovered parameters (:= ground truth included in the posterior credible interval). However, the degree was more severe for population level parameter (hyperparameter), compared to unit level parameters
<img width="1228" alt="image" src="https://user-images.githubusercontent.com/30194633/202731235-5d18d9fb-7823-4ca9-8c41-42d7bc171e9a.png">

**Reply by tomfid** [reply]
> (because it resets)?

Correct. This was an artifact of 32-bit memory limits that should be removed.

**Reply by tomfid** [reply]
> For hierarchical model, none of the team (city, world, service) 

Is the model/data in one of the repos?

**Reply by hyunjimoon** [reply]
I tried to upload models in MonthlyModel folder in each subfolders. e.g. City is [here](https://github.com/Data4DM/BayesSD/tree/master/MonthlyModel/7%20mng(city_world%F0%9F%8C%8F)), starting with 879_urban.

**Reply by tomfid** [reply]
Prospects for autodiff are not looking good, because it requires parallel computation of every equation's gradient, so you can chain rule all the way through the simulation. It's attractive to have, because it would also make the Kalman filter an order of magnitude faster, but it's a very big job.

An alternative might be stochastic HMC (SHMC) but I think the efficiency loss is significant. I think it might also be possible to add some gradient-based proposals to DREAM, which might be a lot like the idea of running a lot of short HMC chains.

**Reply by tomfid** [reply]
> For hierarchical model, none of the team (city, world, service) perfectly recovered parameters (:= ground truth included in the posterior credible interval). 

Was this an adversarial test, or did they know the truth?

**Reply by hyunjimoon** [reply]
Alban (urban model) shared his model which includes subscripts and .vpd. Will mail this to you. Thanks Alban!

**Reply by tomfid** [reply]
Thanks!

**Reply by hyunjimoon** [reply]
Does the following NN flow-based optimization method for ode system look promising to you? @tomfid 
https://github.com/stefanradev93/BayesFlow/blob/master/docs/source/tutorial_notebooks/Linear%20ODE%20system.ipynb

**Comment by hyunjimoon** [comment]
Hazhir advised community might need persuasion that HMC adds value, so we discussed comparing "HMC one gradient computation time" vs "Vensim one simulation" might be useful, but I need to inspect i) how to get one gradient time ii) would it be uniform across all warmup iii) what calculation time would be the most valid comparison.

I wonder what benchmark is used for MCMC algorithms in general. https://mc-stan.org/docs/reference-manual/hmc.html would be the best initialization.

**Comment by hyunjimoon** [comment]
Differentiating the variability from before and after convergence with the variability from the target (posterior distribution) itself is important. 

For next Vensim MCMC, default choices for the end of burn in period should be taken good care. Currently, it has some problem (e.g. After burnin period, vensim resets statistics (e.g. psrf))
